SRVLOGIC-75: Update openshift knative docs with Serverless Logic Workflow guides for 1.32.0 (#89)
openshift-knative/docs: modules/serverless-logic/pages/cloud/build-and-deploy-with-serverless-operator-on-kubernetes.adoc

> = Building and deploying a {context} application on Kubenetes using the {product_name} Serverless Operator

Reviewer: ```suggestion
= (Community Only) Building and deploying a {context} application on Kubenetes using the {product_name} Serverless Operator
```

> = Building and deploying a {context} application on Kubenetes using the {product_name} Serverless Operator

Reviewer 2: ```suggestion
= Building and deploying a {context} application on Kubernetes using the {product_name} Operator
```

:compat-mode!:
// Metadata:
:description: Build and deploy using the Kogito Serverless Workflow Operator a serverless workflow application
:keywords: kogito, workflow, serverless, operator, kubernetes, minikube, openshift
// links
:kogito_serverless_operator_url: https://github.com/kiegroup/kogito-serverless-operator/
:kogito_greeting_example_url: https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-greeting-quarkus
:kaniko_issue_url: https://github.com/GoogleContainerTools/kaniko/issues/2201

> This document describes how to build and deploy your workflow application using a Kubernetes cluster with the link:{kogito_serverless_operator_url}[{product_name} Serverless Operator]. If you don't have any cluster operational, link:{minikube_url}[Minikube] commands are given throughout the guide.

Reviewer: ```suggestion
[WARN]
====
This feature is meant for a quick preview on community repositories, and it's not part of the Developer Preview.
====

This document describes how to build and deploy your workflow application using a Kubernetes cluster with the link:{kogito_serverless_operator_url}[{product_name} Serverless Operator]. If you don't have any cluster operational, link:{minikube_url}[Minikube] commands are given throughout the guide.
```

@kaldesai can you please elaborate better this `WARN`?

cc @radtriste

Reviewer 3: My suggestion would be: Instead of a `warning` can we convert it into a `note`? But the rest of the content looks good to me. 

[NOTE]
====
This feature is meant for a quick preview on community repositories, and it's not part of the developer preview.
====

> This document describes how to build and deploy your workflow application using a Kubernetes cluster with the link:{kogito_serverless_operator_url}[{product_name} Serverless Operator]. If you don't have any cluster operational, link:{minikube_url}[Minikube] commands are given throughout the guide.

Reviewer 4: This one is not resolved ...

Author: Done!

Using the {product_name} Serverless Operator, you will be able to build and deploy a {product_name} {context} application only by having a workflow definition.

If you already have a container built and pushed to a container registry and you want to deploy it on the Kubernetes cluster, then you can do it without the operator following the guide xref:cloud/deploying-on-kubernetes.adoc[Deploying your {context} application on Kubernetes].

The link:{kogito_serverless_operator_url}[{product_name} Serverless Operator] is currently in Alpha version, is under active development and is at the moment supporting {context} definitions that are using:

* Functions
* States
    - Switch including dataConditions
    - Inject including data with a transition
* Operations including Actions containing functionRef with arguments
* KeepActive
* AutoRetries
* ExpressionsLang (jq or jsonpath)

.Prerequisites
* A workflow definition.

> * A Kubernetes cluster with admin privileges 
> +If you haven't got one prepare, you can use a local link:{minikube_url}[Minikube] instance.
> * `kubectl` command-line tool is installed. 
> +Otherwise, Minikube provides it.

Reviewer 2: ```suggestion
* A Kubernetes cluster with admin privileges. If you haven't got one prepare, you can use a local link:{minikube_url}[Minikube] instance.
* `kubectl` command-line tool is installed. Otherwise, Minikube provides it.
```

----

Once the operator is running, it will watch for new custom resources (CR) so that you can prepare your environment to be ready to build a new {context} application based on the definitions you will send to the operator.

== Preparing for the build

You should follow these steps to create a container that you can deploy as a service on Kubernetes.

=== Create a namespace for the building phase

Let's create a new namespace that will hold all the resources that we (or the operator) will create (pods, deployments, services, secretes, config map, and Custom Resources) in this guide.

.Create a namespace for the application to build & run in
[source,bash,subs="attributes+"]
----
kubectl create namespace kogito-workflows
----

=== Create a secret for the container registry authentication
.Create a secret for the container registry authentication
[source,bash,subs="attributes+"]
----
kubectl create secret docker-registry regcred --docker-server=<registry_url> --docker-username=<registry_username> --docker-password=<registry_password> --docker-email=<registry_email> -n kogito-workflows
----

> or you can directly import your local docker config into your the Kubernetes cluster:

Reviewer 2: ```suggestion
or you can directly import your local docker config into your Kubernetes cluster:
```

.Create a secret for the container registry authentication based on local docker config
[source,bash,subs="attributes+"]
----
kubectl create secret generic regcred --from-file=.dockerconfigjson=${HOME}/.docker/config.json --type=kubernetes.io/dockerconfigjson -n kogito-workflows
----

=== Create a {product_name} Serverless Platform containing the configuration (i.e. registry address, secret) for building your workflows

The {product_name} Serverless Platform CR is the resource used to control the behavior of the {product_name} Serverless Operator.
It defines the behavior of all Custom Resources (Workflow and Build) in the given namespace.

Since the {product_name} Serverless Operator is installed in global mode, you will need to specify a {product_name} Serverless Platform CR in each namespace where you want the operator to be executed.

> You can find a basic {product_name} Serverless Platform CR example in the `config/samples` folder that you can simply apply to configure your operator.

Reviewer 4: that one ?

[source,bash,subs="attributes+"]
----
kubectl apply -f https://raw.githubusercontent.com/kiegroup/kogito-serverless-operator/{operator_version}/config/samples/sw.kogito_v1alpha08_kogitoserverlessplatform.yaml -n kogito-workflows
----

Note: In this Custom Resource, spec.platform.registry.secret is the name of the secret you created just before.

[TIP]
====
You can also update "on-the-fly" the {product_name} Serverless Platform CR registry field with this command (change <YOUR_REGISTRY>)

.Create a {product_name} Serverless Platform CR with a specific registry
[source,bash,subs="attributes+"]
----
curl https://raw.githubusercontent.com/kiegroup/kogito-serverless-operator/{operator_version}/config/samples/sw.kogito_v1alpha08_kogitoserverlessplatform.yaml | sed "s|address: .*|address: <YOUR_REGISTRY>" | kubectl apply -f -
----

In order to retrieve the Cluster IP address of Minikube's internal registry to configure your platform, you can use the following command:

.Retrieve Minikube registry internal IP
[source,bash,subs="attributes+"]
----
kubectl get svc registry -n kube-system -ojsonpath='{.spec.clusterIP}'
----
====

== Build and deploy your {context} application

> You can now send your {product_name} {context} Custom Resource to the operator which includes the {context} definition.

Reviewer 2: ```suggestion
You can now send your {product_name} Custom Resource to the operator which includes the {context} definition.
```

Reviewer 2: Maybe we should have another attribute with product_name + context as its value so we don't have to do it every time we create prod docs.  Then we would need to edit just one value in the prod docs.
